core/src/avm1/activation.rs

```diff
@@ -922,10 +924,20 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
     fn action_define_local_2(&mut self) -> Result<FrameControl<'gc>, Error<'gc>> {
         // If the property does not exist on the local object's prototype chain, it is created on the local object.
         // Otherwise, the property is unchanged.
+        // Though this isn't in the SWF19 spec, dot paths and slash paths are also supported and affect the related
+        // object in the same way as Action::SetVariable, if the variable doesn't already exist on the mentioned object.
         let name_val = self.context.avm1.pop();
         let name = name_val.coerce_to_string(self)?;
-        let scope = self.scope_cell();
-        if !scope.read().locals().has_property(self, name) {
+        let mut already_exists = true;
+        if name.find(b":.".as_ref()).is_some() {
+            let value: Value<'gc> = self.get_variable(name)?.into();
+            if value == Value::Undefined {
+                already_exists = false;
+            }
+        } else if !self.scope_cell().read().locals().has_property(self, name) {
+            already_exists = false;
+        }
```

Comment:
We need to differentiate between "not defined" and "defined to undefined", i.e.:

```
function getFoo() {
	trace("getFoo");
	return undefined;
}

function setFoo() {
	trace("setFoo");
}

mc.addProperty("foo", getFoo, setFoo);

trace(mc.foo);
var /mc:foo;
trace(mc.foo);
```
should not call the setter in DefineLocal2 because the property exists, but currently does with this PR.
[Sample SWF](https://github.com/ruffle-rs/ruffle/files/8395023/definelocal2-getset.zip)

I'm also wondering if we can unify the ":." check, since that could happen twice, both here and in `define_local`.

Long term we really want an [Entry](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html)-like API that lets us query everything about a path/var in a single call (whether it exists, value, get/set, etc.). `Callable` is almost like that, but not quite.

In the mean time I'll think a little more about how to do this cleanly with our current situation -- I'll try to get back to you tomorrow!

Reply:
Oh, good catch (again!). 👍 I initially tried to find a way to make this differentiation but wasn't able to find a way to do it. In the end I though it didn't matter - now I know I was wrong. :)